commit 7c68a00cce5b7769c2e34fb7611d14825283fcda
Author: NitzanMordhai <nmordech@ibm.com>
Date:   Wed Jan 21 13:07:56 2026 +0000

    mgr/PyFormatter: prevent segfault on non-printable characters
    
    If PyFormatter::dump_string attempting to format non-printable characters
    can result in a nullptr being passed to dump_pyobject, causing a segfault.
    
    It occured in the following case - crash module tried to load the crash entries
    containing matedata with non-printable characters.
    when these were passed to PyFormatter, the resulting crash prevented ceph-mgr
    from starting.
    
    Fix this by attempting to create a python object using PyUnicode_FromString.
    If it returns nullptr (indicating a decoding failure), the code now falls
    back to decoding using Latin1.
    Additionally, add a guard in dump_pyobject to prevent processing nullptr objects,
    providing a secondary layer of protection against segfaults.
    
    Fixes: https://tracker.ceph.com/issues/74379
    Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 0a82230942a..b33fa5eeba5 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -759,7 +759,15 @@ PyObject *ActivePyModules::get_store_prefix(const std::string &module_name,
   PyFormatter f;
   for (auto p = store_cache.lower_bound(global_prefix);
        p != store_cache.end() && p->first.find(global_prefix) == 0; ++p) {
-    f.dump_string(p->first.c_str() + base_prefix.size(), p->second);
+    PyObject *value = PyUnicode_FromStringAndSize(
+      p->second.c_str(), p->second.size());
+    if (!value) {
+      dout(1) << __func__ << " skipping key " << p->first
+              << " due to PyUnicode_FromStringAndSize failure" << dendl;
+      PyErr_Clear();
+      continue;
+    }
+    f.dump_pyobject(p->first.substr(base_prefix.size()).c_str(), value);
   }
   return f.get();
 }
